src/patterns/patterns.py

```diff
@@ -599,3 +599,300 @@ def make_file_developer_df(
+        print("INFO: Creating developer matrix...")
+
+        # Create the files x developers matrix, using the value_column parameter as the values
+        if 'unique_author' not in self.commit_data.columns:   #self.authors_data = df.merge(df2, how='inner', on='author')
+            self.set_unique_authors()
+
+        if my_df.empty:
+            work_df, stats = self.get_time_range_df(time_range, sum=False)
+        else:
+            # TODO -- enable time ranges with user-provided dataframe my_df
+            if locc_metric not in my_df.columns:
+                err('The dataframe you provided to make_file_developer_df() does '
+                    'not contain the required "%s" column"' % locc_metric)
+            work_df = my_df
+
+        if locc_metric not in work_df.select_dtypes(include=['float64', 'int']):
+            err('get_busfactor_data column parameter must be one of %s' % ','.join(work_df.select_dtypes(
+                include=['float64','int']).columns))
+
+        prim_devs = []
+        secon_devs = []
+        primary_dev = sec_devs = 0
+        tot_developers = 0
+
+        #*1
+        d = pd.DataFrame(work_df.groupby(['filepath', 'unique_author'])[locc_metric].sum())
+        d["dev_knowledge"] = 0
+        d.reset_index(level=d.index.names, inplace=True)
+        #*2
+        authors_commits_df = pd.DataFrame(d.groupby(['unique_author'])[locc_metric].sum())
+        authors_commits_df.reset_index(level=authors_commits_df.index.names, inplace=True)
+        tot_developers = len(authors_commits_df.index)
+        primary_X = 0
+        secondary_X = 0
+        if tot_developers != 0:
+            primary_X = 1/tot_developers
+            secondary_X = primary_X/2
+
+        results = authors_commits_df
+
+        # more knowledge is assigned to the developers that modified the file most times
+        if(metric == 'mul-changes-equal'):
+            #copied *1
+            tot_commits_per_file = pd.DataFrame(d.groupby(['filepath'])[locc_metric].sum())
+            tot_commits_per_file.reset_index(level=tot_commits_per_file.index.names, inplace=True)
+
+            it = 0              #iterator for tot_commits_per_file dataframe
+            for ind in d.index:
+                path = d['filepath'][ind]
+                author = d['unique_author'][ind]
+                d_commits = d[locc_metric][ind]
+                if(path == tot_commits_per_file['filepath'][it]):
+                    tot_commits = tot_commits_per_file[locc_metric][it]
+                    d.iat[ind, d.columns.get_loc('dev_knowledge')] = d_commits/tot_commits
+                else:
+                    it = it+1
+                    tot_commits = tot_commits_per_file[locc_metric][it]
+                    d.iat[ind, d.columns.get_loc('dev_knowledge')] = d_commits/tot_commits
+
+            # display(d.head(5))
+            # display(tot_commits_per_file.head(5))
```

Comment:
Remove commented out lines. 

I also don't think we should have copied code blocks. Can you extract in a separate method? There are some more instances of that below. It's dangerous to have replicated logic since someone can change one place but not the other.

Reply:
Removed the commented lines and added changes to the code to calculate bus factor on specific directories. This eliminated copied code blocks (if any). The copied comments were for my own understanding and memory related to some metrics. Let me know if you come across more of these and I will look into that.